### PR TITLE
[GSoC] Highway network

### PR DESCRIPTION
**Pull request by 17minutes**
This PR is part of my GSoC project "Augmented RNNs".

Implementing a stub for Highway network class. When the time comes: can you give me directions on what to do next to make it work?

### CODE REVIEW

**Comment by zoq** [comment]
Could you take a look into the travis build issue and the style checks?

File: src/mlpack/core/optimizers/sgd/sgd_impl.hpp [hunk 1]
```diff
@@ -107,6 +107,9 @@ double SGD<UpdatePolicyType>::Optimize(
     else
       function.Gradient(iterate, currentFunction, gradient);
 
+    // Time to do something really dirty.
+    // gradient.transform( [](double val) { return std::min(std::max(val, -3.), 3.); } );
```
Comment (zoq) [inline src/mlpack/core/optimizers/sgd/sgd_impl.hpp]: I think we can remove this one here, once gradient clipping is merged.

File: src/mlpack/methods/ann/CMakeLists.txt [hunk 2]
```diff
@@ -21,3 +21,4 @@ add_subdirectory(activation_functions)
 add_subdirectory(init_rules)
 add_subdirectory(layer)
 add_subdirectory(convolution_rules)
+add_subdirectory(augmented)
```
Comment (zoq) [inline src/mlpack/methods/ann/CMakeLists.txt]: Can you clean this PR, and remove all unrelated files?

**Comment by zoq** [comment]
Do you like to finish this first before continuing with the HAM implementation? Both options are fine for me.

**Comment by partobs-mdp** [comment]
> Do you like to finish this first before continuing with the HAM implementation? Both options are fine for me.

Not sure if this is OK, but I would do it in this order:

- Resolve benchmarking PR (since it's almost finished);
- Implement and merge HAM (since this one is the project focus, not the highway network);
- *Resolve this PR*.

Is it OK or should I do it in some other order?

**Comment by zoq** [comment]
@partobs-mdp  Sounds good for me, let's do this.

**Comment by sourabhvarshney111** [comment]
Looks like this is pending from GSoC, 2017. @zoq Can you tell me the issue with this PR? I might be able to do the remaining task.

**Comment by zoq** [comment]
@sourabhvarshney111 sorry for the slow response, this PR contains the augmented related code as well, so the first step would be to separate the Highway network from the rest and see that the test are passing.

**Comment by sourabhvarshney111** [comment]
@zoq Does the augmented rnn code is merged already?
